### DISCUSSION TITLE
Oracle 19C compatibility

### DISCUSSION THREAD

**Discussion by roggss** [nextgenhealthcare/connect#5886] [Q&A]
Hi,

I am not sure if this is still an issue in the current version, but this is an issue of my version 3.9.

A fresh install of 3.9 using Oracle 19c will not create or initialize the DB objects.

Though I am able to recreate the DB objects using the SQL Scripts provided on the page for Oracle and was able to launch Mirth.

However, i found another issue during channel deployment wherein the metadata got deleted.. 

Similar on this thread below. I can paste additional info about the error if needed.

But first, i need to know if this version is compatible with 19c, if not what version at least I should go?

Thanks

https://forums.mirthproject.io/forum/mirth-connect/support/17948-custom-metadata-source-and-type-missing-in-database

UPDATE:
I confirm with version 4.4.. SQL scripts are still not initialized for Oracle 19c.

**Comment by jonbartels** [comment]
This problem sounded familiar. I searched and found a thread from OP in 2022 :( https://github.com/nextgenhealthcare/connect/discussions/5427

Are you seeing the same errors and symptoms for the latest versions?

**Reply by roggss** [reply]
That's my original thread actually a while ago. Though, i found a work around on it, I am still in doubt/fear about migrating our Mirth prod to Oracle 19c because of the second issue that I encounter (the link i provided).

The issue though is not appearing after my last 2 fresh installs, but certainly it pops up to me couple of times on my prior installations. 
This makes me uncomfortable committing this to prod.

Wish, I could see some whitepapers/someone can confirm that 3.9 is compatible with Oracle 19c,

PS: the DB initialization is not working with Mirth 3.9 and 4.4 for Oracle 19c.,

**Reply by ntamll** [reply]
any fix on this issue? is it because Oracle 19c support requires NextGen Connect Enterprise edition instead of Community Edition?

**Reply by pacmano1** [reply]
@ntamll no on the community edition question.  Mirth also does not work that way - Nextgen sells licenses for premium plugins, the base software is the same for all regardless of plugins licensed.

**Reply by jonbartels** [reply]
@ntamll You can learn about how Mirth initializes its databases in [this code](https://github.com/nextgenhealthcare/connect/blob/010f2a4dc636ad1c24b20a7fd1f625c5f6b91a39/server/src/com/mirth/connect/server/migration/ServerMigrator.java#L241) . What is important here is the `throw` statement. Working back up the call stack, if that migration throws there will be a log statement "Failed to migrate database schema".

You can find the SQL files that are used to init a new DB in [this path](https://github.com/nextgenhealthcare/connect/blob/010f2a4dc636ad1c24b20a7fd1f625c5f6b91a39/server/dbconf/oracle)

If you are having issues ONLY with things like channel metadata, [here is the MyBatis script used to create per-channel tables on save/deploy](https://github.com/nextgenhealthcare/connect/blob/010f2a4dc636ad1c24b20a7fd1f625c5f6b91a39/donkey/donkeydbconf/oracle.xml#L73) . This script is referenced in the [JDBC DAO](https://github.com/nextgenhealthcare/connect/blob/010f2a4dc636ad1c24b20a7fd1f625c5f6b91a39/donkey/src/main/java/com/mirth/connect/donkey/server/data/jdbc/JdbcDao.java)

What you need to do is:
* Get the `mirth.log` entries that happen when Mirth is run against a new/clean/empty Oracle DB. Is there a migration failed message?
* Get the `mirth.log` entries right after creating, saving, and deploying a new channel. Are there any logged errors?
* Turn on Oracle logging to capture any DDL statements issued by Mirth
* (If practical) providing [an MCVE](https://stackoverflow.com/help/minimal-reproducible-example) for Oracle may help. Is it practical to reproduce this issue using something like an Oracle Database running in a container? That can help users who do not use Oracle see your issue and apply our Mirth experience to the problem

This information will help the community offer you advice and support. My general guess is that SOMEWHERE there is a clear log entry with a syntax error or warning that a CREATE call is failing or permissions are wrong.

If you need more support than the community can offer (such as whitepapers, specific SLAs, etc.) then engaging consultancies or NextGen would be a better option.
